libraries/kotlin.test/jvm/src/main/kotlin/AssertionsImpl.kt

```diff
@@ -28,8 +28,10 @@ private fun <T : Throwable> assertFailsWithImpl(exceptionClass: Class<T>, messag
         block()
     } catch (e: Throwable) {
         if (exceptionClass.isInstance(e)) {
-            @Suppress("UNCHECKED_CAST")
-            return e as T
+            if (message == null || message == e.message) {
```

Comment:
This should be replaced with a proper assertion and failure. If the message doesn't match you'll get a cryptic failure saying "Expected an exception of type RuntimeException to be thrown, but was RuntimeException".